### PR TITLE
First version of utils

### PR DESCRIPTION
**Pull request by leventdem**
This version includes all the crypto related functions:
- key import/export
- AES enc/dec
- key derivation (for login)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    "type": "git",
+    "url": "git+https://github.com/QwantResearch/masq-common.git"
+  },
+  "keywords": [
+    "encryption",
+    "common",
+    "utils"
+  ],
+  "author": "Levent DEMIR",
+  "license": "ISC",
```
Comment (G-Ray) [inline package.json]: We use MIT in masq-app, at least for now, IDK if we should use MIT here too ?

File: src/crypto.js [hunk 2]
```diff
@@ -0,0 +1,266 @@
+import { generateError, ERRORS, checkObject } from './errors'
+
+const genRandomBuffer = (len = 16) => {
```
Comment (G-Ray) [inline src/crypto.js:3]: This does not return a buffer right ?

File: test/crypto.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
+/* eslint-env mocha */
+/* global MasqCommon */
+/* global chai */
+
+describe('MasqCommon utils', function () {
+  context('Generating a random buffer (e.g. iv)', () => {
+    it('Should generate a random buffer without length parameter', () => {
```
Comment (G-Ray) [inline test/crypto.js:7]: It does not seem to return a buffer, but an array of random values

File: package.json (hunk 1 above)
Comment (leventdem) [reply package.json]: ok done

File: src/crypto.js (hunk 2 above)
Comment (leventdem) [reply src/crypto.js:3]: Right, it returns a int8Array, we can use Buffer.from(returned Array from genRandomBuffer) if necessary.

File: src/crypto.js [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+/**
+ * Generate a PBKDF2 derived key based on user given passPhrase
+ *
+ * @param {string | arrayBuffer} passPhrase The passphrase that is used to derive the key
+ * @param {arrayBuffer} [salt] The salt
+ * @param {Number} [iterations] The iterations number
+ * @param {string} [hash] The hash function used for derivation
+ * @returns {Promise<Uint8Array>}   A promise that contains the derived key
+ */
+const deriveBits = (passPhrase, salt, iterations, hash) => {
```
Comment (G-Ray) [inline src/crypto.js]: should we call this function derivePassPhrase instead ?

File: test/crypto.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  context('Should derive and hash a passphrase ', () => {
+    let passphrase = 'mySecretPass'
+    it('Should derive a passphrase [string] with default settings', async () => {
+      const hashedPassphrase = await MasqCommon.crypto.derivePassphrase(passphrase)
+      chai.assert.equal(hashedPassphrase.hashAlgo, 'SHA-256', 'Default hash algo is SHA-256')
+      chai.assert.equal(hashedPassphrase.iterations, 100000, 'Default iteration is 100000')
+      chai.assert.lengthOf(hashedPassphrase.storedHash, 64, 'SHA-256 returns a 256 bits array, 64 bytes as hex string')
+      chai.assert.lengthOf(hashedPassphrase.salt, 32, 'Default salt is 128 bits array, 32 bytes as hex string')
+    })
+    it('Should reject if passphrase is not a string or is empty', async () => {
```
Comment (G-Ray) [inline test/crypto.js:29]: can you add a newline before each context, describe, it please ?

File: test/utils.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+/* global MasqCommon */
+/* global chai */
+
+describe('MasqCommon utils', () => {
+  context('IndexedDB operations', () => {
+    it('DB should exist', async () => {
+      window.indexedDB.open('test1')
+      const exists = await MasqCommon.utils.dbExists('test1')
+      chai.assert.equal(exists, true, 'database must exist')
+    })
```
Comment (G-Ray) [inline test/utils.js:11]: newline between each test case please

File: test/utils.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+/* eslint-env mocha */
+/* global MasqCommon */
+/* global chai */
+
+describe('MasqCommon utils', () => {
+  context('IndexedDB operations', () => {
```
Comment (G-Ray) [inline test/utils.js:6]: ok

File: test/crypto.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
+/* eslint-env mocha */
+/* global MasqCommon */
+/* global chai */
+
+describe('MasqCommon utils', function () {
```
Comment (G-Ray) [inline test/crypto.js]: masq-common crypto instead

File: test/utils.js (hunk 7 above)
Comment (leventdem) [reply test/utils.js:6]: done

File: test/utils.js (hunk 6 above)
Comment (leventdem) [reply test/utils.js:11]: done

File: test/crypto.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+      } catch (error) {
+        err = error.name
+      }
+      chai.assert.equal(err, MasqCommon.errors.ERRORS.NOPASSPHRASE, 'Reject if passphrase is not a string')
+    })
+
+    it('Should return true if the given passphrase is the same as the stored one', async () => {
+      const hashedPassphrase = await MasqCommon.crypto.derivePassphrase(passphrase)
+      const res = await MasqCommon.crypto.checkPassphrase(passphrase, hashedPassphrase)
+      chai.assert.equal(res, true, 'The check operation should return true')
```
Comment (G-Ray) [inline test/crypto.js]: You can just use `chai.assert.isTrue(res)`

File: src/crypto.js [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+  const iterations = 100000
+  hashedPassphrase.salt = Buffer.from(salt).toString('hex')
+  hashedPassphrase.iterations = iterations
+  hashedPassphrase.hashAlgo = 'SHA-256'
+  return deriveBitsAndHash(passPhrase, salt, iterations)
+    .then(hashedValue => {
+      hashedPassphrase.storedHash = Buffer.from(hashedValue).toString('hex')
+      return hashedPassphrase
+    })
+    .catch(err => console.log(err)
```
Comment (G-Ray) [inline src/crypto.js]: Please use console.error in a catch block

File: test/crypto.js (hunk 8 above)
Comment (leventdem) [reply test/crypto.js]: ok

File: src/crypto.js (hunk 10 above)
Comment (leventdem) [reply src/crypto.js]: ok

File: test/crypto.js (hunk 5 above)
Comment (leventdem) [reply test/crypto.js:29]: done.

File: test/crypto.js (hunk 3 above)
Comment (leventdem) [reply test/crypto.js:7]: done, return Buffer.from

File: test/crypto.js (hunk 9 above)
Comment (leventdem) [reply test/crypto.js]: Right. Done.

File: src/crypto.js [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+/**
+ * Decrypt data
+ *
+ * @param {CryptoKey} key - The AES CryptoKey
+ * @param {string | Object} - The data to encrypt
+ * @param {string} [format] - The ciphertext and iv encoding format
+ */
+const decrypt = (key, ciphertext, format = 'hex') => {
+  _checkCryptokey(key)
+  let context = {}
```
Comment (G-Ray) [inline src/crypto.js]: can you define the object properties at the same time the object is created, like this: `const context = { ciphertext: ..., iv:  ...}

File: src/crypto.js [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+ * @param {CryptoKey} key - The AES CryptoKey
+ * @param {string | Object} - The data to encrypt
+ * @param {string} [format] - The ciphertext and iv encoding format
+ * @returns {Object} - The stringified ciphertext object (ciphertext and iv)
+ */
+const encrypt = (key, data, format = 'hex') => {
+  _checkCryptokey(key)
+  let context = {}
+  let cipherContext = {}
+  context.iv = genRandomBuffer(16)
```
Comment (G-Ray) [inline src/crypto.js]: Same thing, please construct the object at the same time it is declared

Comment (leventdem) [reply src/crypto.js]: ok done in last commit

File: src/crypto.js (hunk 11 above)
Comment (leventdem) [reply src/crypto.js]: ok done in last commit

File: src/crypto.js [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+  return Buffer.from(values)
+}
+
+/**
+ *  Return a buffer of an UInt8Array
+ *
+ * @param {Uint8Array} arr
+ * @returns {Buffer}
+ */
+const getBuffer = (arr) => {
```
Comment (sujeebant) [inline src/crypto.js:19]: Why not use Buffer.from directly ?

File: src/errors.js [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+   * @param {string} name - The name of the error
+   * @param {string} message
+   * @param {string} status
+   */
+const generateError = (name, message, status) => {
+  let err = {
+    name: name || '',
+    message: message || '',
+    status: status || '',
+    error: true
```
Comment (sujeebant) [inline src/errors.js]: Why do we need this attribute ?

File: dist/crypto.js [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+ * @param {string} [hash] The hash function used for derivation
+ * @returns {Promise<Uint8Array>}   A promise that contains the derived key
+ */
+var deriveBits = function deriveBits(passPhrase, salt, iterations, hash) {
+  // Always specify a strong salt
+  if (iterations < 10000) {
+    console.warn('The iteration number is less than 10000, increase it !');
+  }
+
+  return window.crypto.subtle.importKey('raw', typeof passPhrase === 'string' ? Buffer.from(passPhrase) : passPhrase, 'PBKDF2', false, ['deriveBits', 'deriveKey']).then(function (baseKey) {
```
Comment (G-Ray) [inline dist/crypto.js]: Please compute the arg passed to importKey before passing it to the function

File: src/crypto.js [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+/**
+ * Check a given passphrase by comparing it to the stored hash value (in HashedPassphrase object)
+ *
+ * @param {string} passphrase The passphrase
+ * @param {HashedPassphrase} hashedPassphrase The HashedPassphrase object
+ * @returns {Promise<Boolean>}   A promise
+ */
+const checkPassphrase = (passPhrase, hashedPassphrase) => {
+  _checkPassphrase(passPhrase)
+  checkObject(hashedPassphrase, requiredParameterHashedPassphrase)
```
Comment (sujeebant) [inline src/crypto.js:136]: Just a detail, but you could declare requiredParameterHashedPassphrase closer to the checkPassphrase function ?

File: src/crypto.js [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+
+/**
+ * Encrypt data
+ *
+ * @param {CryptoKey} key - The AES CryptoKey
+ * @param {string | Object} - The data to encrypt
+ * @param {string} [format] - The ciphertext and iv encoding format
+ * @returns {Object} - The stringified ciphertext object (ciphertext and iv)
+ */
+const encrypt = (key, data, format = 'hex') => {
```
Comment (sujeebant) [inline src/crypto.js]: Keep the same order between key and data parameters in encrypt and encryptBuffer function

File: dist/crypto.js [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+ */
+var checkPassphrase = function checkPassphrase(passPhrase, hashedPassphrase) {
+  _checkPassphrase(passPhrase);
+  (0, _errors.checkObject)(hashedPassphrase, requiredParameterHashedPassphrase);
+  var salt = hashedPassphrase.salt,
+      iterations = hashedPassphrase.iterations,
+      storedHash = hashedPassphrase.storedHash,
+      hashAlgo = hashedPassphrase.hashAlgo;
+
+  return deriveBitsAndHash(passPhrase, Buffer.from(salt, 'hex'), iterations, hashAlgo).then(function (hashedValue) {
```
Comment (G-Ray) [inline dist/crypto.js]: Can you indent all the `then` to the next line please

File: src/errors.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+/**
+    * @typedef MasqError
+    * @type {Object}
+    * @property {boolean} error
+    * @property {int} status - THe status of the error
+    * @property {string} message - The status of the error
+    * @property {string} name - The name of the error
+    * @property {string} stack - The stack
```
Comment (sujeebant) [inline src/errors.js]: When is stack used ? we should/could use Error object to get stack

File: dist/crypto.js (hunk 18 above)
Comment (leventdem) [reply dist/crypto.js]: Src is ok, this is the dist file.

File: src/errors.js [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+   * Generate a masqError object, the name must be
+   * one of ERRORS.
+   *
+   * @param {string} name - The name of the error
+   * @param {string} message
+   * @param {string} status
+   */
+const generateError = (name, message, status) => {
```
Comment (sujeebant) [inline src/errors.js:62]: Should not be able to generate empty error, or return a default error ?

**Comment by sujeebant** [comment]
Can't we use async functions ?

File: src/crypto.js (hunk 13 above)
Comment (leventdem) [reply src/crypto.js:19]: I add a comment. Buffer.from is not available in test file (without babelify). This function is used only in a regression test.

File: src/crypto.js (hunk 16 above)
Comment (leventdem) [reply src/crypto.js:136]: ok, done

File: src/crypto.js (hunk 17 above)
Comment (leventdem) [reply src/crypto.js]: ok done.

File: src/errors.js (hunk 19 above)
Comment (leventdem) [reply src/errors.js]: Stack is not used, I delete it for now.

File: dist/crypto.js [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+/**
+ * Encrypt data
+ *
+ * @param {CryptoKey} key - The AES CryptoKey
+ * @param {string | Object} - The data to encrypt
+ * @param {string} [format] - The ciphertext and iv encoding format
+ * @returns {Object} - The stringified ciphertext object (ciphertext and iv)
+ */
+var encrypt = function encrypt(key, data) {
+  var format = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'hex';
```
Comment (sujeebant) [inline dist/crypto.js]: Why not add format as a parameter of the encrypt funciton ?

File: src/errors.js (hunk 14 above)
Comment (leventdem) [reply src/errors.js]: I do not remember :-), deletion done.

File: dist/crypto.js (hunk 21 above)
Comment (G-Ray) [reply dist/crypto.js]: Beware, its a dist file ahah

Comment (leventdem) [reply dist/crypto.js]: base64 or hex encoding of iv and ciphertext.

File: src/errors.js (hunk 20 above)
Comment (leventdem) [reply src/errors.js:62]: I add a default name : Unknown error, the message property could be empty. Remove the two others.

File: dist/crypto.js (hunk 21 above)
Comment (sujeebant) [reply dist/crypto.js]: :man_facepalming:

**Comment by leventdem** [comment]
> Can't we use async functions ?



> Can't we use async functions ?

Done.

**Review by sujeebant — APPROVED** [review]

**Pull request merged**